lib/TH/generic/THTensorMath.c

```diff
@@ -1754,22 +1754,61 @@ void THTensor_(topk)(THTensor *rt_, THLongTensor *ri_, THTensor *t, long k, int
   long sliceSize = THTensor_(size)(t, dim);
   THArgCheck(k > 0 && k <= sliceSize, 2, "k not in range for dimension");
 
-  /* Just implement in terms of sort and narrow for now */
-  THTensor* tmpResults = THTensor_(new)();
-  THLongTensor* tmpIndices = THLongTensor_new();
+  THTensor *tmpResults = THTensor_(new)();
+  THTensor_(resize1d)(tmpResults, sliceSize);
+  real *tmp__data = THTensor_(data)(tmpResults);
 
-  THLongStorage* topKSize = THTensor_(newSizeOf)(t);
+  THLongTensor *tmpIndices = THLongTensor_new();
+  THLongTensor_resize1d(tmpIndices, sliceSize);
+  long *tmpi__data = THLongTensor_data(tmpIndices);
+
+  THLongStorage *topKSize = THTensor_(newSizeOf)(t);
   THLongStorage_set(topKSize, dim, k);
   THTensor_(resize)(rt_, topKSize, NULL);
   THLongTensor_resize(ri_, topKSize, NULL);
   THLongStorage_free(topKSize);
 
-  THTensor_(sort)(tmpResults, tmpIndices, t, dim, dir);
-  THTensor_(narrow)(tmpResults, NULL, dim, 0, k);
-  THLongTensor_narrow(tmpIndices, NULL, dim, 0, k);
+  if (dir) {
```

Comment:
> Wouldn't it be simpler to do it with only one macro?

That's true that is saves some repetitions and keep things readable. Also it should not be critical perf. wise to move this branch at the macro level (?) Let's see what other thinks and I can compact this out as suggested!

> Also is the copy of `sliceSize` necessary, because `quickselect` only works on contiguous Tensors

Well `quickselect` supports a stride (see last parameter). But there is a single stride parameter, so it must be the same for the data and the indices. Plus and more importantly `quickselect` works by rearranging the data (swaps) so we need a copy here (this is what `kthvalue` does BTW). Can you think about something better?